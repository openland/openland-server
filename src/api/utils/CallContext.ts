export class CallContext {
    uid?: number;
    oid?: number;
    accountId: number = 0;
    cache: Map<string, any> = new Map<string, any>();
}