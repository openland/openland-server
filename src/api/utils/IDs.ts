import { ID } from '../../modules/ID';

export const IDs = {
    State: new ID('State'),
    County: new ID('County'),
    City: new ID('City'),
    User: new ID('User'),
    Deal: new ID('Deal'),
    Block: new ID('Block'),
    SuperAccount: new ID('SuperAccount'),
    SuperCity: new ID('SuperCity'),
    FeatureFlag: new ID('FeatureFlag'),
    Opportunities: new ID('Op'),
    DebugReader: new ID('dr'),
    Folder: new ID('f'),
    FolderItem: new ID('fi'),
    Task: new ID('task'),
};