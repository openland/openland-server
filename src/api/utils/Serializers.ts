import { Geometry, Polygon } from '../../modules/geometry';

function convertPolygon(polygon: Polygon) {
    let res = [];
    res.push(polygon.coordinates.map((c) => [c.longitude, c.latitude]));
    if (polygon.inner) {
        for (let p of polygon.inner) {
            res.push(p.map((c) => [c.longitude, c.latitude]));
        }
    }
    return res;
}

export function serializeGeometry(geometry?: Geometry | null): string | null {
    if (!geometry) {
        return null;
    }

    return JSON.stringify(geometry.polygons.map(convertPolygon));
}