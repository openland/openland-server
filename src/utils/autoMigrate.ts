import { DefineAttributes, DefineOptions, Model } from 'sequelize';
import { connection } from '../modules/sequelizeConnector';

export function defineModel<TInstance, TAttributes>(
    modelName: string,
    attributes: DefineAttributes,
    options?: DefineOptions<TInstance>
): Model<TInstance, TAttributes> {
    let model = connection.define<TInstance, TAttributes>(modelName, attributes, options) as any; // some type problems here
    // getSchema(model);

    return model;
}

export async function getSchema(model: Model<any, any>) {
    let schema = await connection.query('select * from INFORMATION_SCHEMA.COLUMNS where TABLE_NAME= ? order by ordinal_position;', { replacements: [model.getTableName()]});

    genSchema(schema[0]);
}

let TYPES = {
    'integer': 'INTEGER',
    'character varying': 'STRING',
    'boolean': 'BOOLEAN',
    'timestamp with time zone': 'DATE',
    'json': 'JSON',
};

function genSchema(schema: { column_name: string, data_type: string }[]) {
    let out = '{\n';

    for (let column of schema) {
        let typeName = (TYPES as any)[column.data_type] || 'unknown';

        out += `{ type: sequelize.${typeName}, unique: true },\n`;
    }

    out += '}';

    console.log(out);
}