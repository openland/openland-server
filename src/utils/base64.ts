function unescape(str: string) {
    return (str + '==='.slice((str.length + 3) % 4))
        .replace(/-/g, '+')
        .replace(/_/g, '/');
}

function escape(str: string) {
    return str.replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=/g, '');
}

export function encode(str: string, encoding?: string) {
    return escape(Buffer.from(str, encoding || 'utf8').toString('base64'));
}

export function encodeBuffer(buffer: Buffer, encoding?: string) {
    return escape(buffer.toString('base64'));
}

export function decode(str: string, encoding?: string) {
    return Buffer.from(unescape(str), 'base64').toString(encoding || 'utf8');
}