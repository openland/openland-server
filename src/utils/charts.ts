export interface Chart {
    labels: string[];
    datasets: ChartData[];
}

export interface ChartData {
    label: string;
    values: number[];
}

export interface HistogramRecord {
    count: number;
    value: number;
}

export function reformatHistogram(src: HistogramRecord[], maxBuckets: number = 15): HistogramRecord[] {
    if (src.length === 0) {
        return [];
    }
    let min = src[0].value;
    let max = src[0].value;
    let bucketCount = Math.min(maxBuckets, src.length);
    for (let s of src) {
        if (s.value < min) {
            min = s.value;
        }
        if (s.value > max) {
            max = s.value;
        }
    }

    let buckets: HistogramRecord[] = [];
    let step = (max - min) / bucketCount;
    for (let i = 0; i < bucketCount; i++) {
        let start = min + step * i;
        let end = start + step;
        let v = 0;
        for (let s of src) {
            if (s.value >= start && s.value < end) {
                v += s.count;
            }
        }
        buckets.push({count: v, value: Math.floor(start)});
    }
    return buckets;
}