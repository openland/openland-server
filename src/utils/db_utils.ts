import * as sequelize from 'sequelize';
import { connection } from '../connector';
import { Transaction } from 'sequelize';
import { DB } from '../tables';

export interface Applied<T> {
    id: number;
    created: boolean;
    changed: boolean;
    oldValue?: T;
    newValue?: T;
    changedFields?: string[];
}

export async function findAllRaw<TInstance>(sql: string, model: sequelize.Model<TInstance, any>, tx?: Transaction): Promise<TInstance[]> {
    return (await connection.query(sql, {model: model, raw: true, logging: false, transaction: tx})) as TInstance[];
}

async function findAllTuplesWithNull<TInstance>(tx: Transaction, accountId: number, fields: string[], nullField: string, tuples: any[][], model: sequelize.Model<TInstance, any>): Promise<TInstance[]> {
    let attributes = (model as any).attributes;
    let sqlFields = '(' + fields.map((p) => {
        let attr = attributes[p];
        if (!attr) {
            throw 'Attribute ' + p + ' not found';
        }
        return '"' + p + '"';
    }).join() + ')';
    let sqlTuples = '(' + tuples.map((p) =>
        '(' + p.map((v) => {
            if (v == null || v === undefined) {
                console.warn(p);
                throw 'Null value found!';
            } else if (typeof v === 'string') {
                return connection.escape(v);
            } else {
                return v;
            }
        }).join() + ')'
    ).join() + ')';
    let query = 'SELECT * from "' + model.getTableName() + '" ' +
        'WHERE "account" = ' + accountId + 'AND ' + nullField + ' IS NULL AND ' +
        sqlFields + ' in ' + sqlTuples;
    return findAllRaw(query, model, tx);
}

async function findAllTuplesWithNotNull<TInstance>(tx: Transaction, accountId: number, fields: string[], tuples: any[][], model: sequelize.Model<TInstance, any>): Promise<TInstance[]> {
    let attributes = (model as any).attributes;

    let sqlFields = '(' + fields.map((p) => {
        let attr = attributes[p];
        if (!attr) {
            throw 'Attribute ' + p + ' not found';
        }
        return '"' + p + '"';
    }).join() + ')';
    let sqlTuples = '(' + tuples.map((p) =>
        '(' + p.map((v) => {
            if (v == null || v === undefined) {
                console.warn(p);
                throw 'Null value found!';
            } else if (typeof v === 'string') {
                return connection.escape(v);
            } else {
                return v;
            }
        }).join() + ')'
    ).join() + ')';
    let query = 'SELECT * from "' + model.getTableName() + '" ' +
        'WHERE "account" = ' + accountId + ' AND ' +
        sqlFields + ' in ' + sqlTuples;
    return findAllRaw(query, model, tx);
}

export async function findAllTuples<TInstance>(tx: Transaction, accountId: number, fields: string[], tuples: any[][], model: sequelize.Model<TInstance, any>): Promise<TInstance[]> {
    let attributes = (model as any).attributes;
    let nullable = fields.filter((p) => (attributes[p].allowNull && attributes[p].type.constructor.name === 'STRING'));
    if (nullable.length >= 2) {
        throw 'More than one nullable is not supported!';
    } else if (nullable.length === 1) {
        let withNulls = Array<Array<any>>();
        let withoutNulls = Array<Array<any>>();
        let notNullFields = fields.filter((p) => p !== nullable[0]);
        for (let t of tuples) {
            if (t.filter((p2: any) => p2 === null || p2 === undefined).length > 0) {
                withNulls.push(t.filter((p2: any) => p2 != null && p2 !== undefined));
            } else {
                withoutNulls.push(t);
            }
        }

        if (withNulls.length === 0) {
            return findAllTuplesWithNotNull(tx, accountId, fields, tuples, model);
        } else if (withoutNulls.length === 0) {
            return findAllTuplesWithNull(tx, accountId, notNullFields, nullable[0], withNulls, model);
        } else {
            let notNulledValues = await findAllTuplesWithNotNull(tx, accountId, fields, withoutNulls, model);
            let nulledValues = await findAllTuplesWithNull(tx, accountId, notNullFields, nullable[0], withNulls, model);
            return [...notNulledValues, ...nulledValues];
        }
    } else {
        return findAllTuplesWithNotNull(tx, accountId, fields, tuples, model);
    }
}

export async function bulkAssociations(tx: Transaction, table: string, key1: string, key2: string, values: { value1: number, value2: number }[]) {
    let date = new Date().toUTCString();
    let sqlValues = values.map((v) => `('${date}','${date}',${v.value1},${v.value2})`).join();
    let query = 'INSERT INTO \"' + table + '\" ("createdAt","updatedAt",\"' + key1 + '\",\"' + key2 + '\") VALUES ' + sqlValues + ' ON CONFLICT DO NOTHING';
    await connection.query(query, {logging: false, transaction: tx});
}

export async function bulkInsert<TRow>(tx: Transaction, model: sequelize.Model<any, TRow> | string, rows: TRow[], options?: { inlcudeDates?: boolean }): Promise<number[]> {
    let includeDates = options ? options.inlcudeDates ? options.inlcudeDates : true : true;
    if (includeDates) {
        let date = new Date().toUTCString();
        rows = rows.map(p => ({...(p as any), createdAt: date, updatedAt: date}));
    }
    let table = (typeof (model) === 'string') ? model : model.getTableName() as string;
    let res = await connection.getQueryInterface().bulkInsert(table, rows, {
        raw: true,
        returning: true,
        logging: true,
        transaction: tx
    }) as { id: number }[];
    return res.map(p => p.id);
}

export function escape(value: any): string {
    if (value) {
        if (value === null) {
            return 'NULL';
        } else if (typeof (value) === 'string') {
            return connection.escape(value);
        } else {
            return '' + value;
        }
    } else {
        return 'NULL';
    }
}

// function loadFields<TRow>(src: TRow, fields: string[]) {
//     return `(${fields.map(p => escape((src as any)[p])).join()})`
// }

function loadField<TRow>(src: TRow, field: string) {
    return escape((src as any)[field]);
}

function loadFieldValue<TRow>(src: TRow, field: string) {
    return (src as any)[field];
}

function saveFieldValue<TRow>(src: TRow, field: string, value: any) {
    (src as any)[field] = value;
}

function valueEquals(a: any, b: any) {
    if ((a === undefined || a == null) && (b === undefined || b == null)) {
        return true;
    } else if ((a === undefined || a == null) && (b !== undefined && b != null)) {
        return false;
    } else if ((b === undefined || b == null) && (a !== undefined && a != null)) {
        return false;
    } else {
        return a === b;
    }
}

export async function bulkApply<TRow extends { id?: number, account?: number }>(tx: Transaction, model: sequelize.Model<any, TRow>, accountId: number, key: string, rows: TRow[]) {
    let query = `SELECT * from ${model.getTableName()} WHERE "account" = ${accountId} AND "${key}" IN (${rows.map(r => `${loadField(r, key)}`).join()})`;
    let existing = (await connection.query(query, {transaction: tx, logging: false}))[1].rows as TRow[];
    let forInsert: TRow[] = [];
    let forUpdate: PromiseLike<any>[] = [];
    let indexes = new Array<Applied<TRow>>(rows.length);
    let pendingIndexes: number[] = [];
    let map: { [key: string]: TRow } = {};
    for (let p of existing) {
        map[loadFieldValue(p, key)] = p;
    }
    let index = 0;
    for (let row of rows) {
        let ex = map[loadFieldValue(row, key)];
        let names = Object.getOwnPropertyNames(row);
        if (ex) {
            let updated: TRow = {} as TRow;
            let wasChanged = false;
            let changed = [];
            for (let n of names) {
                if (n === key || n === 'account') {
                    continue;
                }
                let v = loadFieldValue(row, n);
                let v2 = loadFieldValue(ex, n);

                if ((v !== undefined) && !valueEquals(v, v2)) {
                    console.warn('Changed ' + n);
                    console.warn(v2 + ' -> ' + v);
                    saveFieldValue(updated, n, v);
                    wasChanged = true;
                    changed.push(n);
                }
            }
            if (wasChanged) {
                forUpdate.push(model.update(updated, {
                    where: {
                        id: ex.id!!
                    },
                    transaction: tx
                }));
                indexes[index] = {
                    id: ex.id!!,
                    changed: true,
                    created: false,
                    oldValue: ex,
                    newValue: row,
                    changedFields: changed
                };
            } else {
                indexes[index] = {
                    id: ex.id!!,
                    changed: false,
                    created: false
                };
            }
        } else {
            forInsert.push({...(row as any), account: accountId});
            pendingIndexes.push(index);
        }
        index++;
    }

    if (forInsert.length > 0) {
        index = 0;
        for (let ind of await bulkInsert(tx, model, forInsert)) {
            indexes[pendingIndexes[index]] = {
                id: ind,
                created: true,
                changed: false
            };
            index++;
        }
    }

    if (forUpdate.length > 0) {
        for (let p of forUpdate) {
            await p;
        }
    }
    return indexes;
}

export function textLikeFields(query: string, fields: string[]) {
    if (fields.length === 1) {
        return textLikeField(query, fields[0]);
    } else {
        return DB.connection.or(...fields.map(p => textLikeField(query, p)));
    }
}

export function textLikeFieldsText(query: string, fields: string[]) {
    if (fields.length === 1) {
        return textLikeFieldText(query, fields[0]);
    } else {
        return '(' + fields.map(p => textLikeFieldText(query, p)).join(' OR ') + ')';
    }
}

export function textLikeField(query: string, field: string) {
    query = query.toLowerCase().trim()
        .replace('%', '[%]')
        .replace('[', '[[]')
        .replace('\'', '[\']');
    let sq = DB.connection;
    let column = sq.fn('lower', sq.col(field));
    return sq.or(
        sq.where(column, {
            $like: query + '%'
        }),
        sq.where(column, {
            $like: '% ' + query + '%'
        })
    );
}

export function textLikeFieldText(query: string, field: string) {
    query = query.toLowerCase().trim()
        .replace('%', '[%]')
        .replace('[', '[[]')
        .replace('\'', '[\']');

    return '(lower(\"' + field + '\") LIKE \'' + query + '%\' OR lower(\"' + field + '\") LIKE \'% ' + query + '%\')';
}

export async function sumRaw(table: string, field: string, where: string | null): Promise<number> {
    let q = 'SELECT SUM(' + field + ') FROM \"' + table + '\"' + (where ? ' WHERE ' + where : '');
    console.warn(q);
    return (await DB.connection.query(q, {type: DB.connection.QueryTypes.SELECT}))[0].sum || 0;
}

export async function countRaw(table: string, where: string | null): Promise<number> {
    let q = 'SELECT COUNT(*) FROM \"' + table + '\"' + (where ? ' WHERE ' + where : '');
    console.warn(q);
    return (await DB.connection.query(q, {type: DB.connection.QueryTypes.SELECT}))[0].count || 0;
}

export async function percentileRaw(table: string, percentiles: [number], order: string, where: string | null): Promise<number[]> {
    let q = 'SELECT percentile_disc(array[' + percentiles.join() + '])  WITHIN GROUP (ORDER BY ' + order + ') FROM \"' + table + '\"' + (where ? ' WHERE ' + where : '');
    console.warn(q);
    let r = (await DB.connection.query(q, {type: DB.connection.QueryTypes.SELECT}))[0].percentile_disc;
    if (!r) {
        r = [];
    }
    console.warn(r);
    return r;
}

export async function histogramCountRaw(table: string, order: string, where: string | null): Promise<{ count: number, value: number }[]> {
    let q = 'SELECT count(*) as "count", ' + order + ' as "value" FROM \"' + table + '\"' + (where ? ' WHERE ' + where : '') + ' GROUP BY ' + order + ' ORDER BY ' + order;
    console.warn(q);
    let r = (await DB.connection.query(q, {type: DB.connection.QueryTypes.SELECT})) as { count: string, value: number }[];
    if (!r) {
        r = [];
    }
    console.warn(r);
    return r.map((v) => ({count: parseInt(v.count, 10), value: v.value}));
}

export async function histogramSumRaw(table: string, order: string, field: string, where: string | null): Promise<{ count: number, value: number }[]> {
    let q = 'SELECT sum("' + field + '") as "count", ' + order + ' as "value" FROM \"' + table + '\"' + (where ? ' WHERE ' + where : '') + ' GROUP BY ' + order + ' ORDER BY ' + order;
    console.warn(q);
    let r = (await DB.connection.query(q, {type: DB.connection.QueryTypes.SELECT})) as { count: string, value: number }[];
    if (!r) {
        r = [];
    }
    console.warn(r);
    return r.map((v) => ({count: parseInt(v.count, 10), value: v.value}));
}