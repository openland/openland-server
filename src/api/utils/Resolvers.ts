import { CallContext } from '../CallContext';
import { Repos } from '../../repositories';
import { AccessDeniedError } from '../../errors/AccessDeniedError';
import { ErrorText } from '../../errors/ErrorText';

async function fetchPermissions(context: CallContext) {
    if (context.cache.has('permissions')) {
        return (await context.cache.get('permissions')) as Set<string>;
    }
    let res = Repos.Permissions.resolvePermissions({ uid: context.uid, oid: context.oid });
    context.cache.set('permissions', res);
    return await res;
}

async function fetchOrganizationId(context: CallContext) {
    // if (context.cache.has('org_id')) {
    //     return (await context.cache.get('org_id')) as number | null;
    // }
    // let res = DB.User.findById(context.uid).then((v) => v ? v.organizationId as number | null : null);
    // context.cache.set('org_id', res);
    return context.oid !== undefined ? context.oid : null;
}

export function withPermission<T = {}>(permission: string | string[], resolver: (args: T, context: CallContext) => any) {
    return async function (_: any, args: T, context: CallContext) {
        let permissions = await fetchPermissions(context);
        if (Array.isArray(permission)) {
            for (let p of permission) {
                if (permissions.has(p)) {
                    return resolver(args, context);
                }
            }
        } else if (permissions.has(permission)) {
            return resolver(args, context);
        } else {
            throw new AccessDeniedError(ErrorText.permissionDenied);
        }
    };
}

export function withPermissionOptional<T = {}, C = {}>(permission: string | string[], resolver: (args: T, context: CallContext, src: C) => any) {
    return async function (c: C, args: T, context: CallContext) {
        let permissions = await fetchPermissions(context);
        if (Array.isArray(permission)) {
            for (let p of permission) {
                if (permissions.has(p)) {
                    return resolver(args, context, c);
                }
            }
        } else if (permissions.has(permission)) {
            return resolver(args, context, c);
        } else {
            return null;
        }
    };
}

export function withAuth<T = {}>(resolver: (args: T, uid: number) => any) {
    return async function (_: any, args: T, context: CallContext) {
        if (!context.uid) {
            throw new AccessDeniedError(ErrorText.permissionDenied);
        }
        return resolver(args, context.uid!!);
    };
}

export function withAccount<T = {}>(resolver: (args: T, uid: number, org: number) => any) {
    return async function (_: any, args: T, context: CallContext) {
        if (!context.uid) {
            throw new AccessDeniedError(ErrorText.permissionDenied);
        }
        let res = await fetchOrganizationId(context);
        if (res === null) {
            throw new AccessDeniedError(ErrorText.permissionDenied);
        }

        return resolver(args, context.uid!!, res);
    };
}

export function withUser<T = {}>(resolver: (args: T, uid: number) => any) {
    return async function (_: any, args: T, context: CallContext) {
        if (!context.uid) {
            throw new AccessDeniedError(ErrorText.permissionDenied);
        }
        return resolver(args, context.uid!!);
    };
}

export function withAccountTypeOptional<T = {}>(resolver: (args: T, uid?: number, org?: number) => any) {
    return async function (args: T, _: any, context: CallContext) {
        let uid = context.uid;
        let org: number | undefined = undefined;
        if (context.uid) {
            let res = await fetchOrganizationId(context);
            if (res) {
                org = res;
            }
        }
        return resolver(args, uid, org);
    };
}

export function withAny<T = {}>(resolver: (args: T, context: CallContext) => any) {
    return async function (_: any, args: T, context: CallContext) {
        return resolver(args, context);
    };
}