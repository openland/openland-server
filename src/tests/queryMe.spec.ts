import gql from 'graphql-tag';
import prepareApp from './utils/prepareApp';
import { createApiClient } from './utils/createClient';

beforeAll(prepareApp);

const emptyClient = createApiClient();
const defaultClient = createApiClient({ uid: 2 });

describe('Me Query', () => {
    it('Should return null for non-authenticated user', async () => {
        let res = await emptyClient.query({
            query: gql`
              query {
                  me {
                      id
                  }
              }
            `
        });
        expect(res.data).toMatchSnapshot();
    });
    it('Should return profile for user with profile', async () => {
        let res = await defaultClient.query({
            query: gql`
              query {
                  me {
                      id
                      name
                      firstName
                      lastName
                  }
              }
            `
        });
        expect(res.data).toMatchSnapshot();
    });

    it('Should be able to update profile', async () => {
        let res = await defaultClient.mutate({
            mutation: gql`
                mutation {
                    updateProfile(input:{firstName:"Gelb", lastName:"Putintsev"}) {
                        id
                        firstName
                        lastName
                    }
                }
            `
        });
        expect(res.data).toMatchSnapshot();
        let res2 = await defaultClient.query({
            query: gql`
              query {
                  me {
                      id
                      name
                      firstName
                      lastName
                  }
              }
            `
        });
        expect(res2.data).toMatchSnapshot();
        let res3 = await defaultClient.mutate({
            mutation: gql`
                mutation {
                    updateProfile(input:{firstName:"Steve", lastName:"Kite"}) {
                        id
                        firstName
                        lastName
                    }
                }
            `
        });
        expect(res3.data).toMatchSnapshot();
    });

    it('Should preserve old profile info on second createProfile call', async () => {
        let res = await defaultClient.mutate({
            mutation: gql`
                mutation {
                    createProfile(input:{firstName:"Gelb", lastName:"Putintsev"}) {
                        id
                        firstName
                        lastName
                    }
                }
            `
        });
        expect(res.data).toMatchSnapshot();
        let res2 = await defaultClient.query({
            query: gql`
              query {
                  me {
                      id
                      name
                      firstName
                      lastName
                  }
              }
            `
        });
        expect(res2.data).toMatchSnapshot();
    });
});