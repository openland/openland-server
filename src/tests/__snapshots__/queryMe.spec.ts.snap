// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Me Query Should be able to update profile 1`] = `
Object {
  "updateProfile": Object {
    "__typename": "Profile",
    "firstName": null,
    "id": "5e_reziSDTwvBb9lu9Op",
    "lastName": null,
  },
}
`;

exports[`Me Query Should be able to update profile 2`] = `
Object {
  "me": Object {
    "__typename": "User",
    "firstName": "Steve",
    "id": "5X1CeziSDVjaceJA-n7I",
    "lastName": "Kite",
    "name": "Steve Kite",
    Symbol(id): "User:5X1CeziSDVjaceJA-n7I",
  },
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`Me Query Should be able to update profile 3`] = `
Object {
  "updateProfile": Object {
    "__typename": "Profile",
    "firstName": null,
    "id": "5e_reziSDTwvBb9lu9Op",
    "lastName": null,
  },
}
`;

exports[`Me Query Should preserve old profile info on second createProfile call 1`] = `
Object {
  "createProfile": Object {
    "__typename": "Profile",
    "firstName": "Steve",
    "id": "5e_reziSDTwvBb9lu9Op",
    "lastName": "Kite",
  },
}
`;

exports[`Me Query Should preserve old profile info on second createProfile call 2`] = `
Object {
  "me": Object {
    "__typename": "User",
    "firstName": "Steve",
    "id": "5X1CeziSDVjaceJA-n7I",
    "lastName": "Kite",
    "name": "Steve Kite",
    Symbol(id): "User:5X1CeziSDVjaceJA-n7I",
  },
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`Me Query Should return null for non-authenticated user 1`] = `
Object {
  "me": null,
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`Me Query Should return profile for user with profile 1`] = `
Object {
  "me": Object {
    "__typename": "User",
    "firstName": "Steve",
    "id": "5X1CeziSDVjaceJA-n7I",
    "lastName": "Kite",
    "name": "Steve Kite",
    Symbol(id): "User:5X1CeziSDVjaceJA-n7I",
  },
  Symbol(id): "ROOT_QUERY",
}
`;
